ergotree-interpreter/src/eval/savltree.rs

```diff
@@ -0,0 +1,231 @@
+                                        tup.last().clone().try_extract_into::<Vec<u8>>()?,
+                                    );
+                                    tup_items.push((first, second));
+                                } else {
+                                    return Err(EvalError::InvalidResultType);
+                                }
+                            }
+                            _ => unreachable!(),
+                        }
+                    }
+                    tup_items
+                } else {
+                    return Err(EvalError::InvalidResultType);
+                }
+            }
+            _ => return Err(EvalError::InvalidResultType),
+        }
+    };
+
+    let proof = {
+        let v = args.get(1).cloned().ok_or_else(|| {
+            EvalError::NotFound("AvlTree: eval is missing second arg".to_string())
+        })?;
+        Bytes::from(v.try_extract_into::<Vec<u8>>()?)
+    };
+
+    let starting_digest = Bytes::from(avl_tree_data.digest.0.to_vec());
+    let mut bv = BatchAVLVerifier::new(
+        &starting_digest,
+        &proof,
+        AVLTree::new(
+            |digest| Node::LabelOnly(NodeHeader::new(Some(*digest), None)),
+            avl_tree_data.key_length as usize,
+            avl_tree_data.value_length_opt.map(|v| v as usize),
+        ),
+        None,
+        None,
+    )
+    .map_err(map_eval_err)?;
+    for (key, value) in entries {
+        if bv
+            .perform_one_operation(&Operation::Insert(KeyValue { key, value }))
+            .is_err()
+        {
+            return Err(EvalError::Misc(format!(
+                "AvlTree: Incorrect insert for {:?}",
+                avl_tree_data
+            )));
+        }
+    }
+    if let Some(new_digest) = bv.digest() {
+        let digest = ADDigest::sigma_parse_bytes(&new_digest)?;
+        avl_tree_data.digest = digest;
+        Ok(Value::Opt(Box::new(Some(Value::AvlTree(avl_tree_data)))))
+    } else {
+        Err(EvalError::Misc("AvlTree: Cannot update digest".into()))
+    }
+};
+
+fn map_eval_err<T: std::fmt::Debug>(e: T) -> EvalError {
+    EvalError::Misc(format!("{:?}", e))
+}
+
+#[allow(clippy::unwrap_used)]
+#[cfg(test)]
+#[cfg(feature = "arbitrary")]
+mod tests {
+    use std::convert::TryFrom;
+
+    use ergotree_ir::{
+        mir::{
+            avl_tree_data::{AvlTreeData, AvlTreeFlags},
+            constant::{Constant, Literal},
+            expr::Expr,
+            method_call::MethodCall,
+        },
+        types::savltree,
+    };
+    use scorex_crypto_avltree::batch_avl_prover::BatchAVLProver;
+
+    use crate::eval::tests::eval_out_wo_ctx;
+
+    use super::*;
+    #[test]
+    fn eval_avl_tree() {
+        // This example taken from `scorex_crypto_avltree` README
+        let mut prover = BatchAVLProver::new(
+            AVLTree::new(
+                |digest| Node::LabelOnly(NodeHeader::new(Some(*digest), None)),
+                1,
+                None,
```

Comment:
I'm not sure how this parameter is used for AVL operations. Setting it to None here and for `AvlTreeData::value_length_opt` allowed the test to pass.